src/main/webapp/app/iris/base-chatbot/iris-base-chatbot.component.ts

```diff
@@ -0,0 +1,371 @@
+            return;
+        }
+        message.helpful = !!helpful;
+        this.chatService.rateMessage(message, helpful).subscribe();
+    }
+
+    /**
+     * Scrolls the chat body to the bottom.
+     * @param behavior - The scroll behavior.
+     */
+    scrollToBottom(behavior: ScrollBehavior) {
+        setTimeout(() => {
+            const messagesElement: HTMLElement = this.messagesElement.nativeElement;
+            messagesElement.scrollTo({
+                top: 0,
+                behavior: behavior,
+            });
+        });
+    }
+
+    /**
+     * Clear session and start a new conversation.
+     */
+    onClearSession(content: any) {
+        this.modalService.open(content).result.then((result: string) => {
+            if (result === 'confirm') {
+                this.isLoading = false;
+                this.chatService.clearChat();
+            }
+        });
+    }
+
+    /**
+     * Accepts the permission to use the chat widget.
+     */
+    acceptPermission() {
+        this.chatService.setUserAccepted();
+        this.userAccepted = true;
+    }
+
+    /**
+     * This method is intended to handle the closing of the chat interface.
+     * It emits a close event which should be handled by the parent component.
+     */
+    closeChat() {
+        this.chatService.messagesRead();
+        this.closeClicked.emit();
+    }
+
+    /**
+     * Handles the key events in the message textarea.
+     * @param event - The keyboard event.
+     */
+    handleKey(event: KeyboardEvent): void {
+        if (event.key === 'Enter') {
+            if (!this.isLoading && this.active) {
+                if (!event.shiftKey) {
+                    event.preventDefault();
+                    this.onSend();
+                } else {
+                    const textArea = event.target as HTMLTextAreaElement;
+                    const { selectionStart, selectionEnd } = textArea;
+                    const value = textArea.value;
+                    textArea.value = value.slice(0, selectionStart) + value.slice(selectionEnd);
+                    textArea.selectionStart = textArea.selectionEnd = selectionStart + 1;
+                }
+            } else {
+                event.preventDefault();
+            }
+        }
+    }
+
+    /**
+     * Handles the input event in the message textarea.
+     */
+    onInput() {
+        this.adjustTextareaRows();
+    }
+
+    /**
+     * Handles the paste event in the message textarea.
+     */
+    onPaste() {
+        setTimeout(() => {
+            this.adjustTextareaRows();
+        }, 0);
+    }
+
+    /**
+     * Adjusts the height of the message textarea based on its content.
+     */
+    adjustTextareaRows() {
+        const textarea: HTMLTextAreaElement = this.messageTextarea.nativeElement;
+        textarea.style.height = 'auto'; // Reset the height to auto
+        const lineHeight = parseInt(getComputedStyle(textarea).lineHeight, 10) + 4;
```

Comment:
Could we replace the magic numbers with variables here? 
What does `+4` relate to?

I am also unsure why we are using `parseInt(getComputedStyle(textarea).lineHeight, 10)` - I think we should use `Number(...)` instead

Reply:
4 accounts for the spacing between lines, otherwise the last line is only half viewable after 3 lines have been written. 
10 is clear I hope. `Number()` is incompatible because we are parsing e.g. "20px", not a pure number.